### PR TITLE
Feature/photon history

### PR DESCRIPTION
**Pull request by kmtsui**
Toggle WCSim to save photon scattering and reflection history when `-DWCSIM_SAVE_PHOTON_HISTORY_FLAG=ON` is set in `cmake` configuration. This feature is mostly for calibration studies when we want to know the photon propagation histories with respect to different water qualities and surface reflectivity.

Major changes include:
1. Save `trackID` per photon in `WCSimWCHit` and `WCSimWCDigi`
2. Save the number of scattering and reflection experienced by a photon in `WCSimTrajectory`. In `WCSimTrajectory::AppendStep()`, check whether a Rayleigh/Mie scattering happens, or whether a boundary reflection occurs and save the boundary surface type (`kOtherS,kBlackSheetS,kReflectorS, kPhotocathodeS` defined in `WCSimEnumerations`)
3. Construct new `WCSimRootCherenkovHitHistory` class to save the photon scattering and reflection histories. Users can access this class in a way similar to `WCSimRootCherenkovHitTime`, thus get the scattering and reflection histories per Cherenkov photon. An example usage is also documented in `sample-root-scripts/sample_readfile.C`.
4. In `WCSimEventAction::EndOfEventAction()`, retrieve photon histories from `WCSimTrajectory` container and save in output root file.
5. The photon process check and photon histories saving only occur when `-DWCSIM_SAVE_PHOTON_HISTORY_FLAG=ON` is set in `cmake` configuration. This will enforce `WCSimTrajectory` to be stored in `WCSimTrackingAction::fpTrackingManager` for every photon, overriding setup from `/Tracking/fractionOpticalPhotonsToDraw`.

Bug fix:
1. In `WCSimWCTriggerBase::AlgNDigits`, if the digi hit time range is smaller than trigger window, the find trigger loop is invalid. Pushed a fix to enforce `window_start_time<=window_end_time` in all case

### CODE REVIEW

File: sample-root-scripts/sample_readfile.C
Comment (tdealtry) [inline sample-root-scripts/sample_readfile.C]: Maybe nicer to use `EnumAsString()` here?
More verbose, but clearer

File: include/WCSimRootEvent.hh
Comment (tdealtry) [inline include/WCSimRootEvent.hh]: Does this return the parent Cherenkov photon? The number of scatters this Cherenkov photon has experienced? Something else?
Need a comment here to clarify, and potentially a more informative method name

File: include/WCSimRootEvent.hh
Comment (tdealtry) [inline include/WCSimRootEvent.hh]: Surely this is what we want?
```suggestion
  std::vector<ReflectionSurface_t> fReflec;
```
and similarly elsewhere in this class

File: include/WCSimRootEvent.hh
Comment (tdealtry) [inline include/WCSimRootEvent.hh]: Could call it `GetReflectionSurfaces()`
(at the very least, it should be plural)

File: include/WCSimTrajectory.hh
Comment (tdealtry) [inline include/WCSimTrajectory.hh]: Why not use `ReflectionSurface_t` for these functions? Similarly for the member variables

File: src/WCSimTrajectory.cc
Comment (tdealtry) [inline src/WCSimTrajectory.cc]: Naive question, but would it ever be useful to know how many of each scattering process (rather than the combined total)?
Now would be the perfect time to split the numbers up

File: include/WCSimWCHit.hh
Comment (tdealtry) [inline include/WCSimWCHit.hh]: Please rename to something like `AddTrackID()` to make it clear that it's pushing back, rather than setting the only value

**Review by tdealtry — CHANGES_REQUESTED** [review]
Thanks @kmtsui. A few inline comments, mostly about naming of things. Also
- Please add a comment to WCSim.mac to say that `/Tracking/fractionOpticalPhotonsToDraw` is ignored when `-DWCSIM_SAVE_PHOTON_HISTORY_FLAG=ON` is set (and 100% of photons are saved).
- The CI runs successfully, which is good news! I would like to see something comparing results with `-DWCSIM_SAVE_PHOTON_HISTORY_FLAG` set to ON and OFF though, to ensure nothing odd is happening

**Comment by tdealtry** [comment]
Good spot on the random number fix. This shows why we should put in more than 1 random number generator in GHOST...

Can you pull in from develop? It looks like your IWCD geom fix isn't accounted for in the CI.

**Comment by kmtsui** [comment]
It seems that photon tracks with WLS in OD PMTs are recorded in the output root files, I am trying to remove them...
And somehow when there are too many tracks, the [CI](https://github.com/WCSim/WCSim/actions/runs/7148770243/job/19470222190?pr=397#step:11:9363) is failing. Any thoughts?

**Comment by tdealtry** [comment]
It's when it's running [the WCSim file -> simpler tree convertor](https://github.com/WCSim/Validation/blob/master/Generate/daq_readfilemain.C) I guess its running out of memory
Looking [here](https://wcsim.github.io/Validation/061d6447e4cc91fc896fed70a7e0a97ec5370f7f/2/2/index.html) we see that it gets through the 20" ok, and fails for mPMTs at ~event 170 of 200. Each call of the convertor does 20" then mPMT then OD, so there's some kind of memory leak going on - I've applied a fix (2 actually - 1] to cleanup some memory at the end of the function, 2] as a more foolproof thing by changing how the script is called - one PMT type per process)

**Comment by kmtsui** [comment]
CI broken due to [typo ](https://github.com/WCSim/Validation/blob/fa5f3ea25f52c06a54f3e4b76cbaf0f75417aa0d/Generate/Generate.sh#L4) `pmtype` <->`pmttype`

**Comment by tdealtry** [comment]
Doh! Fixed and rerunning on your latest commit

**Comment by kmtsui** [comment]
Thanks. The CI checks are finished and looking good. In the end, I need to change the track saving flags such that they are not applied to optical photons. But the side effect is that photon tracks will not be saved at all when we want to use the `ProcessList` or `ParticleList` control. 
I am think whether it is good to roll back to [d3ee4f0](https://github.com/WCSim/WCSim/pull/397/commits/d3ee4f03e997afa53fc2d3afcd1745489940b28d) and accept the large number of additional tracks as known feature.

**Comment by tdealtry** [comment]
All green is always good!

I've lost track a little on precisely what's going on...

- [The CI](https://wcsim.github.io/Validation/061d6447e4cc91fc896fed70a7e0a97ec5370f7f/index.html) on 061d644 (effectively d3ee4f0) is all green for everything but geometries with ODs, where the number of tracks is increasing. In IWCD, the increase a small (up to 20%) but for HK FD it's large (3000%+).
  - Need to grab the file size information from the CI really, to see if this is something to be worrying about. But it's potentially worrying.
- Since the CI is green for all other geometries, this suggests it's extra optical photons that are being saved related to something in the OD. Likely the WLS plates
  - Why is the WLS special? Is it something to do with extra wavelength shifting processes generating two levels of optical photon tracks? (Something like Cherenkov photon -> absorption in WLS (gets a != 0 track ID) -> remmission = new optical photon)
- It seems like `-DSAVE_PHOTON_HISTORY=ON` has been the default for quite a few CI runs. Why then has the tracking information not changed for all detectors? Since (among other things) this effectively sets `/Tracking/fractionOpticalPhotonsToDraw` to 100%

I suppose if we need some slightly different logic for `-DSAVE_PHOTON_HISTORY` = ON and OFF, in order to give the functionality & keep file size in check, we can do. Since the checks do show that things are identical, other than extra OD tracks, when things are ON. We can have OFF be the size light mode, with the fine control with `ProcessList` or `ParticleList` control options

**Comment by kmtsui** [comment]
The large increase in the number of tracks being saved is due to:
1. [All photon trajectories are stored](https://github.com/kmtsui/WCSim/blob/feature/photon_history/src/WCSimTrackingAction.cc#L73) (equivalent to `/Tracking/fractionOpticalPhotonsToDraw 100`) 
2. [Optical photons from WLS are saved ](https://github.com/kmtsui/WCSim/blob/feature/photon_history/src/WCSimTrackingAction.cc#L135)because of the "OpWLS" being added to [ProcessList](https://github.com/kmtsui/WCSim/blob/feature/photon_history/src/WCSimTrackingAction.cc#L29)
3. [Parent photon tracks of the WLS photons are saved](https://github.com/kmtsui/WCSim/blob/feature/photon_history/src/WCSimTrackingAction.cc#L172C35-L172C57) becaue of the `saveHitProducingTracks` flag

Only the OD events have such WLS processes, so we see differences only in those OD geometries. We will see the same behavior if we use the `/Tracking/fractionOpticalPhotonsToDraw 100` command instead of `-DSAVE_PHOTON_HISTORY=ON`

If we want to keep the `/Tracking/fractionOpticalPhotonsToDraw` usage, it will need more an extra flag to mark whether the photon trajectories are coming from `/Tracking/fractionOpticalPhotonsToDraw` or `-DSAVE_PHOTON_HISTORY=ON`, or do another random number throw in the `PostUserTrackingAction()`

**Comment by kmtsui** [comment]
I have added an extra flag to mark whether to save the photon track in ROOT. Now `/Tracking/fractionOpticalPhotonsToDraw` has the same behavior no matter `-DSAVE_PHOTON_HISTORY` is `ON` or `OFF`. Accordingly `ProcessList` and `ParticleList` controls should also work properly.

**Comment by tdealtry** [comment]
Thanks for the explanation & the careful work with this @kmtsui 
I'm happy to merge after you set the default for `WCSIM_SAVE_PHOTON_HISTORY_FLAG` to `OFF`

**Review by tdealtry — APPROVED** [review]

**Comment by tdealtry** [comment]
Thanks @kmtsui

**Comment by gondiaz** [comment]
Some methods here were modified, for example in `WCSimWCHit`, `SetTrackID` which is now `SetTrackID` and `GetTrackID()` to `GetTrackID(int i)`.

This changes directly affect [WCSimFQTuner](https://github.com/fiTQun/WCSimFQTuner), so I would suggest that from now on, before merging anything into `WCSim` we should also port the changes to the tuner (which could also be part of WCSim though). I was not following the details of the new functionality, so could you check for it?

**Comment by kmtsui** [comment]
@gondiaz The change in `WCSimWCHit` and `WCSimWCDigi` is to save all photon track ID (instead of the just first one) for a PMT with multiple photon hit. I don't have write access to [WCSimFQTuner](https://github.com/fiTQun/WCSimFQTuner) so cannot update it.
The fact that WCSimFQTuner lives under another (private) organization makes it difficult to maintain. We would need the admin there to help with the integration.

**Pull request merged**
